src/Native/System.Native/pal_networking.cpp

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_config.h"
+#include "pal_networking.h"
+#include "pal_utilities.h"
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <functional>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <vector>
+
+const int NUM_BYTES_IN_IPV4_ADDRESS = 4;
+const int NUM_BYTES_IN_IPV6_ADDRESS = 16;
+
+static_assert(PAL_HOST_NOT_FOUND == HOST_NOT_FOUND, "");
+static_assert(PAL_TRY_AGAIN == TRY_AGAIN, "");
+static_assert(PAL_NO_RECOVERY == NO_RECOVERY, "");
+static_assert(PAL_NO_DATA == NO_DATA, "");
+static_assert(PAL_NO_ADDRESS == NO_ADDRESS, "");
+
+static void IpStringToAddressHelper(const char* address, const char* port, bool isIpV6, int32_t* err, std::function<void(const addrinfo& info)> lambda)
+{
+    assert(address != nullptr);
+    assert(err != nullptr);
+
+    addrinfo hint = {};
+    hint.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+    hint.ai_family = isIpV6 ? AF_INET6 : AF_INET;
+    hint.ai_socktype = 0;
+    hint.ai_protocol = 0;
+
+    addrinfo* info = nullptr;
+    int result = getaddrinfo(address, port, &hint, &info);
+    if (result == 0)
+    {
+        *err = 0;
+        lambda(*info);
```

Comment:
The lambda just points to more unmanaged code. I contemplated merging the IPv4 and IPv6 code into one function, like the IpAddressToString function, but it would turn into a messy prototype with a multiple optional parameters that would be used depending on the version; due to this, I figured the minuscule overhead of another function call was worth it for the code cleanliness, but if you think the optional parameters would be better then I'd be happy to change it :)